lib/mempool/indexer.js

```diff
@@ -0,0 +1,481 @@
+ *
+ * and orphan related events as they are not part of
+ * the coin or tx indexers as orphans.
+ *
+ *
+ * @alias module:mempool.MempoolIndexer
+ */
+
+class MempoolIndexer {
+  /**
+   * Create a mempool indexer.
+   * @param {Mempool} mempool
+   */
+
+  constructor(options) {
+    this.options = new MempoolIndexerOptions(options);
+    this.mempool = this.options.mempool;
+
+    this.coinIndex = new CoinIndex();
+    this.txIndex = new TXIndex();
+
+    this.init();
+  }
+
+  /**
+   * Start listening for the mempool events
+   */
+
+  init() {
+    this.mempool.on('unconfirmed', (tx, block) => this.unconfirmed(tx, block));
+    this.mempool.on('add entry', (entry, view) => this.addEntry(entry, view));
+    this.mempool.on('remove entry', entry => this.removeEntry(entry));
+  }
+
+  /**
+   * We have new entry in the mempool.
+   * `add entry` is emitted after `tx`.
+   * We can index coins.
+   * - We received it from the network.
+   * - Block was disconnected.
+   * - Orphan got resolved
+   * @param {MempoolEntry} entry
+   * @param {CoinView} view
+   */
+
+  addEntry(entry, view) {
+    const tx = entry.tx;
+
+    this.txIndex.insert(entry, view);
+
+    for (const {prevout} of tx.inputs) {
+      const {hash, index} = prevout;
+
+      this.coinIndex.remove(hash, index);
+    }
+
+    for (let i = 0; i < tx.outputs.length; i++)
+      this.coinIndex.insert(tx, i);
+  }
+
+  /**
+   * Transaction was removed from mempool.
+   * This might happen for several reasons:
+   *  - Mempool size limit got rid of it.
+   *  - Transaction was included in block.
+   *  - Double spend in a block
+   *  - After reorg tx is no longer final
+   * We concentrate on recovering inputs as coins.
+   * @param {MempoolEntry} entry
+   */
+
+  removeEntry(entry) {
+    const tx = entry.tx;
+    const hash = tx.hash();
+
+    this.txIndex.remove(hash);
+
+    for (const {prevout} of tx.inputs) {
+      const {hash, index} = prevout;
+      const prev = this.mempool.getTX(hash);
+
+      if (!prev)
+        continue;
+
+      this.coinIndex.insert(prev, index);
+    }
+
+    for (let i = 0; i < tx.outputs.length; i++)
+      this.coinIndex.remove(hash, i);
+  }
+
+  /**
+   * Block disconnected and we recover coins if they are available.
+   * This event comes after `tx` and `add entry`, we might want to
+   * check if outputs indexed by those are already spent in the mempool.
+   * @param {TX} tx
+   * @param {Block} block
+   */
+
+  unconfirmed(tx, block) {
+    const hash = tx.hash();
+
+    for (let i = 0; i < tx.outputs.length; i++) {
+      if (this.mempool.isSpent(hash, i))
+        this.coinIndex.remove(hash, i);
+    }
+  }
+
+  /**
+   * Reset indexes
+   * @private
+   */
+
+  reset() {
+    this.txIndex.reset();
+    this.coinIndex.reset();
+  }
+
+  /**
+   * Find all transactions pertaining to a certain address.
+   * Note: this does not accept multiple addresses.
+   * @param {Address} addrs
```

Comment:
Why does this function accept `addrs` instead of `addr` and then place the single address into an array to iterate over if it only accepts a single address?